atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampUtils.java

```diff
@@ -0,0 +1,145 @@
+                            CassandraTimestampUtils.ROW_AND_COLUMN_NAME,
+                            "current_max_ts",
+                            ColumnValueDescription.forType(ValueType.FIXED_LONG)))),
+            ConflictHandler.IGNORE_ALL);
+
+    private static final long CASSANDRA_TIMESTAMP = -1;
+
+    private static final String ROW_AND_COLUMN_NAME = "ts";
+    private static final String ROW_AND_COLUMN_NAME_HEX_STRING = encodeCassandraHexString(ROW_AND_COLUMN_NAME);
+
+    private static final String APPLIED_COLUMN = "[applied]";
+    private static final String VALUE_COLUMN = "value";
+
+    private static final byte[] SUCCESSFUL_OPERATION = {1};
+
+    private static final String BACKUP_COLUMN_NAME = "oldTs";
+    private static final byte[] INVALIDATED_VALUE = new byte[1];
+
+    private CassandraTimestampUtils() {
+        // utility class
+    }
+
+    public static ByteBuffer constructSelectTimestampBoundQuery() {
+        String selectQuery = String.format(
+                "SELECT %s FROM %s WHERE key=%s AND column1=%s AND column2=%s;",
+                VALUE_COLUMN,
+                wrapInQuotes(AtlasDbConstants.TIMESTAMP_TABLE.getQualifiedName()),
+                ROW_AND_COLUMN_NAME_HEX_STRING,
+                ROW_AND_COLUMN_NAME_HEX_STRING,
+                CASSANDRA_TIMESTAMP);
+        return ByteBuffer.wrap(PtBytes.toBytes(selectQuery));
+    }
+
+    public static ByteBuffer constructCheckAndSetCqlQuery(Long expected, long target) {
+        String updateQuery = (expected == null)
+                ? constructInsertIfNotExistsQuery(target)
+                : constructUpdateIfEqualQuery(expected, target);
+
+        return ByteBuffer.wrap(PtBytes.toBytes(updateQuery));
+    }
+
+    public static boolean wasOperationApplied(CqlResult cqlResult) {
+        Column appliedColumn = getNamedColumn(cqlResult, APPLIED_COLUMN);
+        return Arrays.equals(appliedColumn.getValue(), SUCCESSFUL_OPERATION);
+    }
+
+    public static long getLongValue(CqlResult cqlResult) {
+        Column valueColumn = getNamedColumn(cqlResult, VALUE_COLUMN);
+        if (Arrays.equals(INVALIDATED_VALUE, valueColumn.getValue())) {
+            throw new IllegalStateException("Couldn't extract a long from the invalidated value!");
+        }
+        return PtBytes.toLong(valueColumn.getValue());
+    }
+
+    private static Column getNamedColumn(CqlResult cqlResult, String columnName) {
+        List<Column> columns = getColumnsFromOnlyRow(cqlResult);
+        return Iterables.getOnlyElement(
+                columns.stream()
+                        .filter(column -> columnName.equals(PtBytes.toString(column.getName())))
+                        .collect(Collectors.toList()));
+    }
+
+    private static String constructUpdateIfEqualQuery(Long expected, long target) {
+        return String.format(
+                "UPDATE %s SET value=%s WHERE key=%s AND column1=%s AND column2=%s IF value=%s;",
```

Comment:
`UPDATE ... IF ...` is [typically how CAS is written in CQL](http://docs.datastax.com/en/cql/3.1/cql/cql_using/use_ltweight_transaction_t.html); CQL doesn't have a CAS operator, as far as I know.